compiler/rustc_parse/src/parser/item.rs

```diff
@@ -59,10 +59,18 @@ impl<'a> Parser<'a> {
 
         let post_attr_lo = self.token.span;
         let mut items = ThinVec::new();
-        while let Some(item) = self.parse_item(ForceCollect::No)? {
-            self.maybe_consume_incorrect_semicolon(Some(&item));
+
+        // There shouldn't be any semicolons before or afer items.
+        // `parse_item` consumes the appropriate semicolons so any leftover is an error.
+        let mut last = None;
+        while let Some(item) = {
+            self.maybe_consume_incorrect_semicolon(last);
+            self.parse_item(ForceCollect::No)?
+        } {
```

Comment:
Using a block in a `while` loop's predicate is not a style I have seen used in the compiler code. Maybe move those statements into the loop body and use a `loop` instead instead of a `while`?

Reply:
Okay, I agree that it can be a bit harder to follow. Done :)